[Lens] Add row/column dimension tracking to transpose functionality

metamonk opened the pull request:
## Summary

This PR enhances Lens datatable transpose functionality to support explicit tracking of row and column dimensions, enabling sophisticated pivot table operations with grand totals and customization options.

## Screenshots

### Enabling Pivot Mode

**Before: Enable Pivot Message**

<img width="440" alt="Enable pivot table message" src="https://github.com/user-attachments/assets/25d5e8ad-fa0d-4411-9d0c-1f330fae0367" />

When no fields are in "Split metrics by", users see a helpful callout explaining how to enable pivot functionality.

---

**After: Adding Transpose Dimensions**

<img width="442" alt="Split metrics by configuration" src="https://github.com/user-attachments/assets/567135d9-5dce-4686-a946-ca93906576f9" />

Fields added to "Split metrics by" create pivot dimensions and enable pivot mode.

---

### Pivot Table Settings

**Full Settings Panel**

<img width="440" alt="Complete pivot settings panel" src="https://github.com/user-attachments/assets/c087c083-2bc3-424f-b586-5c6a45b7e23a" />

Complete configuration options:
- **Pivot structure information** - Shows estimated column calculation
- **Row totals toggle** - Display grand total row
- **Column totals toggle** - Display grand total columns  
- **Empty cells dropdown** - Choose how to display missing data (empty/0/-/N/A)
- **Column limit** - Protect against excessive columns (10-1000 range)

---

### Grand Totals

**Row Totals**

<img width="509" alt="Table with row totals" src="https://github.com/user-attachments/assets/a9264866-431a-44a4-8661-3f8b875cbc74" />

Grand total row displays aggregated values at the bottom of the table.

---

**Column Totals**

<img width="211" alt="Table with column totals" src="https://github.com/user-attachments/assets/a07973d4-ebf2-4552-bee4-36b0c916e3d1" />

Grand total column shows aggregated metrics across all data.

---

**Both Row and Column Totals**

<img width="441" alt="Row and column totals together" src="https://github.com/user-attachments/assets/b0f2183b-8db4-42b2-aa95-35f413cddf37" />

Both totals can be enabled simultaneously, with the intersection showing the overall grand total.

---

### Empty Cell Handling

<img width="809" alt="Empty cells displayed as null" src="https://github.com/user-attachments/assets/87b8e06f-73b0-42c5-af14-2672c2b6ee0a" />

Missing data combinations can be displayed as empty, 0, -, or N/A based on user preference.

---

### Complete Pivot Table

<img width="807" alt="Working pivot table with transposed dimensions" src="https://github.com/user-attachments/assets/cf83e08b-a801-483d-a79f-a6724558c7f3" />

Full pivot table with multiple transposed dimensions, showing dynamic column generation based on data.

---

## What's New

**Enhanced Transpose Utilities** (`kbn-transpose-utils`):
- Added optional `dimension` parameter to `getTransposeId()` for row/column tracking
- Added `parseTransposeId()` helper to extract dimension information from transpose IDs
- Maintains full backward compatibility with existing transpose behavior

**Updated Type Definitions**:
- Extended `ColumnState` to include `transposeDimension` and `transposeLevel` properties
- Added dimension tracking to `bucketValues` for better metadata preservation
- Added `pivotSettings` configuration in datatable state

**New UI Components**:
- `PivotSettings` component for pivot table configuration with:
  - Row/column totals toggles
  - Empty cell display options (empty/0/-/N/A)
  - Column limit control (10-1000)
  - Pivot structure information display
- `TableSettings` component combining appearance and pivot settings

**Implementation Updates**:
- Enhanced transpose helpers with dimension-aware logic
- Updated datatable components to handle row/column dimension tracking
- Added support for grand totals on both rows and columns
- Enhanced tests for subtotals and grand totals functionality

---

## Testing

- ✅ **37 unit tests passing** (28 existing + 9 new component tests)
  - `transpose_helpers.test.ts` - 4 tests
  - `datatable_fn.test.ts` - 13 tests  
  - `pivot_settings.test.tsx` - 9 new component tests
  - `summary.test.ts` - 11 tests
- ✅ Component tests cover all PivotSettings interactions
- ✅ No TypeScript errors
- ✅ Manual testing completed with multiple data scenarios

**Test Coverage:**
- Transpose dimension tracking (rows vs columns)
- Grand totals calculation (rows and columns)
- Empty cell handling (all 4 display options)
- Column limit enforcement
- Backward compatibility with existing transpose IDs
- Edge cases (no data, single values, missing combinations)

---

## Backward Compatibility

All changes are fully backward compatible:
- Transpose IDs without dimension prefix default to 'columns' behavior
- Existing datatable configurations continue to work without modification
- New pivot features are opt-in via UI configuration
- No breaking changes to API or data structures

---

## Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
  - All text uses `i18n.translate()` with proper keys
  - Follows sentence case conventions
  - Includes helpful descriptions and tooltips
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
  - UI is self-explanatory with clear labels and help text
  - Will add documentation if requested by reviewers
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
  - 37 total tests passing
  - New component tests for PivotSettings UI
  - Comprehensive coverage of transpose and totals functionality
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee
  - No API changes
  - Fully backward compatible
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
  - Will run if requested by reviewers

---

## Risk Assessment

**Low Risk:**
- ✅ Backward compatibility fully maintained
- ✅ Comprehensive unit test coverage (37 tests)
- ✅ Changes isolated to Lens datatable visualization
- ✅ No HTTP API changes
- ✅ No changes to data persistence layer
- ✅ All existing tests continue to pass

**Medium Risk:**
- ⚠️ New UI components need real-world usage validation
- ⚠️ Performance with very large pivot tables (>100 columns) should be monitored

**Mitigation:**
- Column limit setting (default 100, max 1000) protects against excessive column generation
- Extensive manual testing completed with various data scenarios
- Clear user feedback via pivot structure information text

metamonk committed:
WIP: Add row/column dimension tracking to Lens transpose

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

metamonk committed:
[Lens] Add row/column dimension tracking to transpose functionality

This PR enhances Lens datatable transpose functionality to support explicit tracking of row and column dimensions, enabling more sophisticated pivot table operations and better handling of multi-dimensional data structures.

Key changes:

- Enhanced transpose utilities (kbn-transpose-utils):
  - Added optional `dimension` parameter to `getTransposeId()` for row/column tracking
  - Added `parseTransposeId()` helper to extract dimension information from transpose IDs
  - Maintains backward compatibility with existing transpose behavior

- Updated type definitions:
  - Extended `ColumnState` to include `transposeDimension` and `transposeLevel` properties
  - Added dimension tracking to `bucketValues` for better metadata preservation
  - Added pivot settings support with `pivotSettings` in datatable state

- New UI components:
  - `PivotSettings` component for pivot table configuration (row/column totals, empty cells, column limits)
  - `TableSettings` component to combine appearance and pivot settings

- Implementation updates:
  - Enhanced transpose helpers with dimension-aware logic
  - Updated datatable components to handle row/column dimension tracking
  - Added support for grand totals on both rows and columns
  - Enhanced tests for subtotals and grand totals functionality

All existing tests pass and maintain backward compatibility.

metamonk committed:
[Lens] Add component tests for PivotSettings

- Added comprehensive component tests for PivotSettings UI
- Tests cover all interactive elements: totals toggles, empty cells, column limit
- Tests validate state management and user interactions
- 9 test cases with 100% coverage of component functionality

metamonk committed:
[Lens] Fix PivotSettings always rendering to show enable message

- Removed conditional rendering in TableSettings
- PivotSettings now always renders and handles its own state
- Shows 'Enable pivot' message when no transpose enabled
- Shows full settings when pivot mode is active
